### PR TITLE
Add LIAM and MeLIBA ego algorithms

### PR DESCRIPTION
**Pull request by FernandezR**
Added implementations for two ego algorithms: LIAM and MeLIBA.

Tested in using the LBF env and a trained FCP partner.

The branch can be deleted once merged.

### CODE REVIEW

File: common/run_episodes.py [hunk 1]
```diff
@@ -35,16 +39,16 @@ def run_single_episode(rng, env, agent_0_param, agent_0_policy,
         avail_actions=avail_actions_0,
         hstate=init_hstate_0,
         rng=act0_rng,
-        aux_obs=None,
+        aux_obs=(init_act_onehot["agent_0"].reshape(1, 1, -1), init_joint_act_onehot, init_reward["agent_0"].reshape(1, 1, -1)),
```
[the remark this answers is not in the document]
Comment (FernandezR) [reply common/run_episodes.py:42]: This is the easiest way to do it at the moment, since there is a centralized run_episodes script. We can try to find a better way to do it later as we enhance the codebase.

File: marl/liam_utils.py
[the remark this answers is not in the document]
Comment (FernandezR) [reply marl/liam_utils.py:1]: The PPO one is in marl, that's why I put it there. Do you think we should be moving that one as well?

File: agents/agent_interface.py [hunk 2]
```diff
@@ -404,25 +404,451 @@ def get_action_value_policy(self, params, obs, done, avail_actions, hstate, rng,
...
         dummy_done = jnp.zeros((1, batch_size))
         dummy_avail = jnp.ones((1, batch_size, self.action_dim))
         dummy_x = (dummy_obs, dummy_done, dummy_avail)
-        
+
         # Initialize model using the pre-initialized network
         return self.network.init(rng, init_hstate, dummy_x)
-    
+
+class LIAMPolicy(AgentPolicy):
```
[the remark this answers is not in the document]
Comment (FernandezR) [reply agents/agent_interface.py:429]: Sure, I can do it that way. The current way just followed the existing framework.

File: agents/agent_interface.py [hunk 3]
```diff
@@ -404,25 +404,451 @@ def get_action_value_policy(self, params, obs, done, avail_actions, hstate, rng,
...
+            hstate=hstate[1],
+            rng=rng,
+            aux_obs=aux_obs,
+            env_state=env_state
+        )
+
+        return action, val, pi, (new_encoder_hstate, new_policy_hstate)
+
+    @partial(jax.jit, static_argnums=(0,))
+    def evaluate(self, params, obs, done, avail_actions, hstate, rng,
```
[the remark this answers is not in the document]
Comment (FernandezR) [reply agents/agent_interface.py:573]: What would you prefer? I pulled this from LIAM  but I can change it to something else.

File: agents/agent_interface.py [hunk 4]
```diff
@@ -404,25 +404,451 @@ def get_action_value_policy(self, params, obs, done, avail_actions, hstate, rng,
...
+            pi: jnp.Array, the policy distribution
+            recon_loss1: jnp.Array, the reconstruction loss from the decoder
+            recon_loss2: jnp.Array, the partner reconstruction loss from the decoder
+            new_hstate: tuple(jnp.Array, jnp.Array), the new hidden state for the encoder and policy
+        """
+        act, _, _ = aux_obs
+
+        embbeding, new_encoder_hstate = self.encoder.compute_embedding(
+            params=params['encoder'],
+            hstate=hstate[0],
```
[the remark this answers is not in the document]
Comment (FernandezR) [reply agents/agent_interface.py:607]: I'll add a comment that specifies the name. Changing the variable name itself will just add more lines of code.

File: agents/initialize_agents.py [hunk 5]
```diff
@@ -2,25 +2,27 @@
...
+
     Returns:
         policy: S5ActorCriticPolicy, the policy object
         params: dict, initial parameters for the agent
     """
     # Create the S5 policy with direct parameters
     policy = S5ActorCriticPolicy(
         action_dim=env.action_space(env.agents[0]).n,
-        obs_dim=env.observation_space(env.agents[0]).shape[0],
+        obs_dim=config.get("POLICY_INPUT_DIM", env.observation_space(env.agents[0]).shape[0]),
```
[the remark this answers is not in the document]
Comment (FernandezR) [reply agents/initialize_agents.py:25]: Already done.

File: ego_agent_training/liam_ego.py [hunk 6]
```diff
@@ -0,0 +1,631 @@
...
+
+Only supports a population of homogeneous RL partner agents.
+
+Warning: modify with caution, as this script is used as the main script for ego training throughout the project.
+
+
+
+Command to run LIAM ego training:
+
+python ego_agent_training/run.py algorithm=liam_ego/lbf task=lbf label=test_liam_ego
```
[the remark this answers is not in the document]
Comment (FernandezR) [reply ego_agent_training/liam_ego.py:14]: Both do work

File: ego_agent_training/liam_ego.py [hunk 7]
```diff
@@ -0,0 +1,631 @@
...
+    Train LIAM ego agent using the given partner checkpoints and initial ego parameters.
+
+    Args:
+        config: dict, config for the training
+        env: gymnasium environment
+        train_rng: jax.random.PRNGKey, random key for training
+        ego_policy: AgentPolicy, policy for the ego agent
+        init_ego_params: dict, initial parameters for the ego agent
+        n_ego_train_seeds: int, number of ego training seeds
+        encoder: Encoder, the encoder model
```
[the remark this answers is not in the document]
Comment (FernandezR) [reply ego_agent_training/liam_ego.py:63]: The capitalization is consistent with the class and variable names

File: common/run_episodes.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (carolinewang01) [reply common/run_episodes.py:42]: Sounds good

File: marl/liam_utils.py
[the remark this answers is not in the document]
Comment (carolinewang01) [reply marl/liam_utils.py:1]: No, please leave that one. The ppo_utils.py is there because our implemented MARL algorithm is ppo. So i didnt want to duplicate code unneccesarily.

File: ego_agent_training/configs/algorithm/meliba_ego/_base_.yaml [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+
+DECODER_STATE_EMBED_DIM: 64
+DECODER_AGENT_CHARACTER_EMBED_DIM: 32
+DECODER_HIDDEN_DIM: 64
+DECODER_OUTPUT_DIM: 32
+
+DECODER_LOSS_COEFF: 1.0
+DECODER_KL_WEIGHT: 0.05
+
+POLICY_INPUT_DIM: 256 # (Latent dim * 4) + observation dimension (obs dim added in the run script)
```
[the remark this answers is not in the document]
Comment (carolinewang01) [reply ego_agent_training/configs/algorithm/meliba_ego/_base_.yaml:37]: I see. Nevertheless it's confusing to have the temp value here for the policy input dim bc it looks incorrect. You can delete this param then from the configs and only specify it within the main ego script.

File: ego_agent_training/liam_ego.py [hunk 9]
```diff
@@ -0,0 +1,631 @@
...
+                    unroll=16,
+                )
+
+                return advantages, advantages + traj_batch.value
+
+            def _update_minbatch(init_state, batch_info):
+                train_state, encoder_decoder_train_state = init_state
+                init_ego_hstate, traj_batch, advantages, returns = batch_info
+                def _loss_fn(params, encoder_decoder_params, init_ego_hstate, traj_batch, gae, target_v):
+
```
[the remark this answers is not in the document]
Comment (carolinewang01) [reply ego_agent_training/liam_ego.py:248]: Place it in a way that makes sense with the policy and value losses. Something that somebody could ctrl-f for

File: agents/agent_interface.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (carolinewang01) [reply agents/agent_interface.py:429]: I understand. We should plan to factor out the others as well. The interface should only only contain interface related codes.

File: agents/agent_interface.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (carolinewang01) [reply agents/agent_interface.py:607]: Sounds good.

**Comment by FernandezR** [comment]
All the corrections have been made. On the heldout eval set MeLIBA > LIAM > PPO on the overall mean.
